R/add_miQC.R

```diff
@@ -34,29 +34,43 @@ add_miQC <- function(sce, posterior_cutoff = 0.75, seed = NULL){
   set.seed(seed)
 
   # generate linear mixture model of probability of cells being compromised
-  model <- tryCatch(
-    miQC::mixtureModel(sce),
-    error = function(x){NA}
-  )
-  if (is(model, "flexmix")  && length(model@components) < 2){
-    # first model fit gave only one component, give it one more chance.
-    model <- miQC::mixtureModel(sce)
+  model <- NULL
+  pass_cells <- NULL
+  model_attempt <- 0
+
+  # This can fail in a few ways, so we will wrap the next steps in a while/try loop
+  while(model_attempt < 3 &&
+        (!is(model, "flexmix") || length(model@components) < 2 ) &&
+        is.null(pass_cells)){
+    model_attempt <- model_attempt + 1
+    try({
+      model <- miQC::mixtureModel(sce)
+      # filter step can fail, so we will try this too, keeping the passing cell ids
+      pass_cells <- colnames(
+        miQC::filterCells(sce,
+                          model = model,
+                          posterior_cutoff = posterior_cutoff,
+                          verbose = FALSE)
+      )
```

Comment:
Can you check the indentation on this? It seems funny here in that line 43 is more indented than line 45 which doesn't seem correct to me?

Reply:
It does look strange, but 43 is indented for the `while` conditions, and 45 is the inside of the `while` block, so they are correct by the rules I was following.